hindsight/io_modules/derived_stream.lua

```diff
@@ -0,0 +1,197 @@
+-- This Source Code Form is subject to the terms of the Mozilla Public
+-- License, v. 2.0. If a copy of the MPL was not distributed with this
+-- file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+local M = {}
+local assert    = assert
+local error     = error
+local pairs     = pairs
+local pcall     = pcall
+local require   = require
+local tonumber  = tonumber
+
+local batch_checkpoint_update   = batch_checkpoint_update
+local read_config               = read_config
+local read_message              = read_message
+
+local io        = require "io"
+local math      = require "math"
+local os        = require "os"
+local string    = require "string"
+
+setfenv(1, M) -- Remove external access to contain everything in the module
+
+local SEC_IN_DAY = 60 * 60 * 24
+
+function load_schema(name, schema)
+    local ticker_interval   = read_config("ticker_interval")
+    local cfg_name          = read_config("cfg_name")
+    local format            = read_config("format")
+    local ts_field          = read_config("ts_field") or "Timestamp"
+    local files             = {} -- manages the derive stream buffer/output files
+
+    if format == "redshift" then
+        local driver        = require "luasql.postgres"
+        local rs            = require "derived_stream.redshift"
+
+        local db_config     = read_config("db_config") or error("db_config must be set")
+        local buffer_path   = read_config("buffer_path") or error("buffer_path must be set")
+        local buffer_size   = read_config("buffer_size") or 10000 * 1024
+        assert(buffer_size > 0, "buffer_size must be greater than zero")
+
+        local last_insert   = 0
+        local uuid          = nil
+
+        local env = assert(driver.postgres())
+        local con = assert(env:connect(db_config.name, db_config.user, db_config._password, db_config.host, db_config.port))
+
+        local function get_output_file(ts)
+            local day = math.floor(ts / (SEC_IN_DAY * 1e9))
+            local f = files[day]
+            if not f then
+                local date = os.date("%Y%m%d", ts / 1e9)
+                local table_name = string.format("%s_%s", name, date)
+                local filename = string.format("%s/%s_%s.sql", buffer_path, cfg_name, date)
+                local ok, cnt, err = pcall(con.execute, con, rs.get_create_table_sql(table_name, schema))
+                if ok and err and not string.match(err, "duplicate key violates unique constraint") then
```

Comment:
Please add a comment that the string match detects concurrent "create table" statements, and that failure is expected.